### PR TITLE
Enhanced CI RPC check

### PR DESCRIPTION
**Pull request by lalexgap**
This enhances our CI by making use of the `start-rpc-servers` script from `go-nitro` and the `create-channels` script from `nitro-gui`.  By using multiple clients and creating multiple channels we increase the chance of catching a bug.

The downside is that spinning up the RPC servers in `start-rpc-servers` can take awhile, resulting in the GH action taking around 5 minutes!

### CODE REVIEW

**Comment by lalexgap** [comment]
Moving this back to draft as I'm not sure the check is working like I think.

File: .github/workflows/rpc-test.yml [hunk 1]
```diff
@@ -28,20 +30,16 @@ jobs:
...
-        run: go run . -naaddress 0x0000000000000000000000000000000000000000 -config ./scripts/test-configs/alice.toml &> output.log &
+      - name: Run go-nitro RPC servers
+        run: go run ./scripts/start-rpc-servers.go &> output.log &
         working-directory: code/go-nitro
 
       - name: Install dependencies
         run: yarn
 
-      - name: Launch Address request using CLI tool and test against expectations
-        run: bash ./test-against-server.sh
```
Comment (geoknee) [inline .github/workflows/rpc-test.yml]: Do we want to remove this file?

Comment (lalexgap) [reply .github/workflows/rpc-test.yml]: Yup, good point!

Comment (lalexgap) [reply .github/workflows/rpc-test.yml]: https://github.com/statechannels/nitro-gui/pull/27/commits/541a7c33babf5227e95f6d50cb9b0b2902d08c58

File: packages/nitro-rpc-client/src/transport/http.ts [hunk 2]
```diff
@@ -17,6 +17,13 @@ export class HttpTransport {
   public static async createTransport(server: string): Promise<Transport> {
     // eslint-disable-next-line new-cap
     const ws = new w3cwebsocket(`ws://${server}/subscribe`, undefined);
+
+    // throw any websocket errors so we don't fail silently
+    ws.onerror = (e) => {
+      console.error("Error with websocket connection to server");
+      throw e;
+    };
```
Comment (geoknee) [inline packages/nitro-rpc-client/src/transport/http.ts:30]: Is this related to this PR?

File: packages/nitro-rpc-client/scripts/client-runner.ts [hunk 3]
```diff
@@ -134,3 +153,47 @@ async function wait(ms: number) {
...
+}
+
+// Checks if the server is up by sending a simple get_address POST request
+// This is specific to the HTTP/WS RPC transport
+async function isServerUp(port: number): Promise<boolean> {
+  let result: AxiosResponse<unknown, unknown>;
+
+  try {
+    const req = generateRequest("get_address", {});
+    result = await axios.post(
```
Comment (geoknee) [inline packages/nitro-rpc-client/scripts/client-runner.ts]: Is it weird that this uses http POST rather than GET?

Comment (geoknee) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: Why not use the rpc client to launch the `get_address` request?

File: .github/workflows/rpc-test.yml [hunk 4]
```diff
@@ -28,21 +30,17 @@ jobs:
...
 
       - name: Install dependencies
         run: yarn
 
-      - name: Launch Address request using CLI tool and test against expectations
-        run: bash ./scripts/test-against-server.sh
-        working-directory: ./packages/nitro-rpc-client
+      - name: Run Create Channels script
+        # TODO: We could write a test specific script that creates channels and checks the results
+        run: npx ts-node ./scripts/client-runner.ts create-channels -w 300000 &> output.log
```
Comment (geoknee) [inline .github/workflows/rpc-test.yml:42]: Under what conditions will this cause the action to fail? Before we were doing a very basic assertion on the output of the CLI tool -- now I guess we will only fail on errors?

File: packages/nitro-rpc-client/scripts/client-runner.ts (hunk 3 above)
Comment (lalexgap) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: > Is it weird that this uses http POST rather than GET?

Our `go-nitro` RPC server **only accepts** POST request, so we have to use POST here.

File: .github/workflows/rpc-test.yml (hunk 4 above)
Comment (lalexgap) [reply .github/workflows/rpc-test.yml:42]: > Under what conditions will this cause the action to fail? Before we were doing a very basic assertion on the output of the CLI tool -- now I guess we will only fail on errors?

We're no longer explicitly checking the result but because we're running the full `create-channels` script I still think we're getting a lot more coverage than we previously had.  Previously we were just checking if the address from the get address call was correct. Now we're creating ledger channels,virtual channels and using them to pay, and if anything breaks down doing that we should report an error.

I think if we do want to start adding assertions in, we should write a test specific script that the `client-runner` runs. It could look like `create-channels` but could also assert on the results after creating the channels.

File: packages/nitro-rpc-client/scripts/client-runner.ts (hunk 3 above)
Comment (lalexgap) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: > Why not use the rpc client to launch the `get_address` request?

We're trying to use the `get_address` query a way to check if the RPC server is available yet, since the RPC servers can take some time to spin up on CI. So basically we're checking if the RPC server is ready to spin up the RPC client. If we try to spin up the RPC client before the RPC server is ready the RPC client will just crash trying to talk to a server that isn't listening.

Comment (geoknee) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: > > Is it weird that this uses http POST rather than GET?
> 
> Our `go-nitro` RPC server **only accepts** POST request, so we have to use POST here.

My question really was, it is weird the server only accepts POST requests?

**Review by geoknee — APPROVED** [review]

File: packages/nitro-rpc-client/scripts/client-runner.ts (hunk 3 above)
Comment (lalexgap) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: > My question really was, it is weird the server only accepts POST requests?

I think it kinda makes sense. The JSON-RPC interface is accessed using one endpoint, and different methods are called by specifying the `method` in your JSON-RPC request data. Even for cases like `get_address` , that do not require any method  parameters,  the consumer is still sending  JSON data in the POST body.  For example in the case of `getAddress` we're posting JSON that looks like this:
```
'{"jsonrpc":"2.0","method":"getAddress","params":[],"id":67}'
```

File: packages/nitro-rpc-client/src/transport/http.ts (hunk 2 above)
Comment (lalexgap) [reply packages/nitro-rpc-client/src/transport/http.ts:30]: Slightly, this was causing the http transport to fail silently when it couldn't make a connection to the websocket. Happy to pull this off into it's own PR.

Comment (lalexgap) [reply packages/nitro-rpc-client/src/transport/http.ts:30]: Moved to it's own PR, https://github.com/statechannels/nitro-gui/pull/54

File: packages/nitro-rpc-client/scripts/client-runner.ts (hunk 3 above)
Comment (geoknee) [reply packages/nitro-rpc-client/scripts/client-runner.ts]: https://www.simple-is-better.org/json-rpc/transport_http.html#get-request
